File: Ej2_Java/src/main/java/org/example/Ejercicios.java
package org.example;

import java.util.*;
import java.io.IOException;
import java.util.Random;



public class Ejercicios {

    Scanner scanner = new Scanner(System.in);  // Crear el objeto Scanner
    Random random = new Random();

    /**
     * 1. Saludo personalizado
     * Pedir al usuario su nombre y saludarle con un mensaje personalizado. Ejemplo: "¡Hola,
     * Juan!".
     */
    public void ejercicio1() {


        System.out.print("Ingresa nombre: ");
        String nombre = scanner.nextLine();

        System.out.println("Hola " + nombre);
    }

    /**
     * 2. Número par o impar
     * Pedir un número al usuario e indicar si es par o impar.
     * @throws IOException
     */
    public void ejercicio2()
    {
        System.out.print("Introduce numero para comprobar par o impar: ");
        Double numero = scanner.nextDouble();
        if (numero % 2 == 0){
            System.out.println("Par");
        }else {
            System.out.println("Impar");
        }
    }

    /**
     * 3. Cálculo del área de un círculo
     * Pedir al usuario el radio de un círculo y calcular su área (A = π * r^2).
     */
    public void ejercicio3(){
        System.out.print("Introduce radio para calcular el area de un circulo: ");

        try{
            Double numero = scanner.nextDouble();
            Double radio = Math.PI * (numero * numero);
            System.out.println("El area de un circulo de " + numero + " de radio es: " + String.format("%.2f", radio));

        }catch (Exception e){
            System.out.println("Debes introducir un valor numerico. Error: " + e);
        }
    }

    /**
     * 4. Generador de números aleatorios
     * Generar un número aleatorio entre 1 y 100 y pedir al usuario que adivine qué número es.
     * El programa indicará si el número es mayor o menor al ingresado.
     * Pista: Utilizar la función de generación de números aleatorios y bucles while o do-while
     */
    public void ejercicio4() {
        int numRandom = random.nextInt(0,100);
        System.out.print("Introduce numero para adivinar: ");
        try{
            int num;
            int cont = 0;

            do {
                num = scanner.nextInt();
                cont++;
                if(num < numRandom){
                    System.out.println("El numero oculto es mayor");
                }else{
                    System.out.println("El numero oculto es menor");
                }

            }while (num != numRandom);

            System.out.println("Bingo!!. Num de intentos: " + cont);

        }catch (Exception e){
            System.out.println("Debes introducir un valor numerico. Error: " + e);
        }
    }

    /**
     * 5. Tabla de multiplicar
     * Pedir un número al usuario e imprimir su tabla de multiplicar del 1 al 10.
     */
    public void ejercicio5() {
        System.out.print("Introduce numero: ");

        try{
            double num = scanner.nextDouble();
            for (int i = 0; i < 10; i++) {
                System.out.println(num + " x " + (i+1) + " --> " + (num * (i+1)));
            }
        }catch (Exception e){
            System.out.println("Debes introducir un valor numerico. Error: " + e);
        }
    }

    /**
     * 6. Sumar números del 1 al 100
     * Crear un programa que sume todos los números del 1 al 100.
     */
    public void ejercicio6() {
        int suma = 0;

        for (int i = 0; i < 100; i++) {
            suma += (i+1);
        }

        System.out.println("Suma de los numeros 1-100: " + suma);
    }


    /**
     * 7. Reverso de una cadena
     * Pedir una cadena al usuario e imprimirla al revés
     */
    public void ejercicio7() {
        System.out.print("Introduce una cadena para darle la vuelta: ");
        String cadena = scanner.nextLine();

        String reversed = "";
        for (int i = cadena.length() - 1; i >= 0; i--) {
            reversed += cadena.charAt(i);
        }

        System.out.println("Cadena reversed: " + reversed);
    }

    /**
     * 8. Contador de vocales
     * Pedir una frase al usuario y contar cuántas vocales (a, e, i, o, u) contiene.
     */
    public void ejercicio8() {
        Map<String, Integer> vocales = new HashMap<>();

        vocales.put("a", 0);
        vocales.put("e", 0);
        vocales.put("i", 0);
        vocales.put("o", 0);
        vocales.put("u", 0);

        System.out.print("Introduce una frase para contar sus vocales: ");
        String frase = scanner.nextLine().toLowerCase();

        for (int i = 0; i < frase.length(); i++) {
            String letra = String.valueOf(frase.charAt(i)); // Obtiene el valor de i

            if (vocales.containsKey(letra)) {
                vocales.put(letra, vocales.get(letra) + 1); //vocales.get(letra) obtiene el valor de la clave letra
            }
        }

        System.out.println("Cantidad de vocales en la frase "+ frase);

        // Recorrer todas las entradas de un dicc
        for(Map.Entry<String, Integer> entrada : vocales.entrySet()){ // Entrada es cada par clave-valor
            System.out.println(entrada.getKey() + " -> " + entrada.getValue());
        }
    }

    /**
     * 9. Números primos
     * Pedir un número al usuario y verificar si es un número primo.
     */
    public void ejercicio9() {
        System.out.print("Introduce número a comprobar primo:");
        int numero = scanner.nextInt();

        if (checkPrimo(numero)) {
            System.out.println("El numero " + numero + " es primo");
        }else{
            System.out.println("El numero " + numero + " no es primo");
        }
    }

    public boolean checkPrimo(int num){

        // Menor o igual a 1 no es primo
        if (num <= 1){
            return false;
        }

        // 2 es primo
        if (num == 2){
            return true;
        }

        // si es mayor que 2 y par, no es primo
        if (num % 2 == 0){
            return false;
        }

        // Ddivisores desde 3 hasta la raíz cuadrada de n
        for (int i = 3; i * i <= num; i += 2) { // Solo numeros impares
            if (num % i == 0) {
                return false; // Si num es divisible por i, no es primo
            }
        }

        return true;
    }

    /**
     * 10. Conversión de temperaturas
     * Crear un programa que convierta una temperatura de grados Celsius a Fahrenheit y
     * viceversa. (C = (F - 32) * 5/9, F = C * 9/5 + 32).
     */
    public void ejercicio10() {

        System.out.println("Elige una opcion:\n1 - De Celsius a Fahrenheit\n2 - De Fahrenheit a Celsius");

        try{
            int opc = scanner.nextInt();

            if (opc != 1 && opc != 2){
                throw new Exception();
            }

            if (opc == 1) celToFahrenheit();
            if (opc == 2) fahrenheitToCelsius();


        }catch (Exception e){
            System.out.println("Debes elegir entre las opciones disponibles");
        }


    }

    public void celToFahrenheit() {
        System.out.print("Introduce grados celsius: ");
        try{
            double cel = scanner.nextDouble();
            double fahrenheit = (cel * 9/5) + 32;
            System.out.println(cel + " Celsius son -> " + String.format("%.2f", fahrenheit) + " Fahrenheit");

        }catch (Exception e){
            System.out.println("Debe introducir un valor numerico. Error: " + e);
        }
    }


    public void fahrenheitToCelsius() {
        System.out.print("Introduce grados fahrenheit: ");
        try{
            double fah = scanner.nextDouble();
            double celsius = (fah - 32) * 5/9;
            System.out.println(fah + " Fahrenheit son -> " + String.format("%.2f",celsius) + " Celsius");

        }catch (Exception e){
            System.out.println("Debe introducir un valor numerico. Error: " + e);
        }
    }

    /**
     * 11. Fibonacci
     * Crear un programa que imprima los primeros N números de la secuencia de Fibonacci (el
     * número N lo ingresa el usuario)
     */
    public void ejercicio11() {

        System.out.print("Introduce cantidad de números de la secuencia de Fibonacci: ");

        try{
            int num = scanner.nextInt();
            int x = 0;
            int y = 1;

            for (int i = 1; i <= num; i++) {
                System.out.println(x + " ");

                int z = x + y; // Suma de los dos numeros para obtener el tercero
                //Intercambiamos los valores del 1 al 2, y del 2 al 3
                x = y;
                y = z;
            }
        }catch (Exception e){
            System.out.println("Debes introducir un valor numerico. Error: " + e);
        }
    }

    /**
     * 12. Inversión de un número
     * Pedir un número entero al usuario e imprimir el número con sus dígitos invertidos.
     * Ejemplo: si el usuario ingresa 1234, el programa debería imprimir 4321.
     */
    public void ejercicio12(){
        System.out.print("Introduce numero a invertir: ");

        try {
            int numero = scanner.nextInt();

            String reversed = "";
            for (int i = String.valueOf(numero).length() - 1; i >= 0; i--) {
                reversed += String.valueOf(numero).charAt(i);
            }

            System.out.println("Numero invertido: " + reversed);
        }catch (Exception e){
            System.out.println("Debe introducir un entero!");
        }
    }

    /**
     * 13. Clase Anime sencilla
     * Crear una clase Anime con los atributos nombre, episodios y género. Luego, permitir al
     * usuario crear un objeto de tipo Anime con estos atributos y mostrar su información.
     */
    public void ejercicio13(){
        System.out.println("Introduce datos del anime: ");
        System.out.print("Nombre: ");
        String nombre = scanner.next();

        System.out.print("Episodios: ");
        String episodios = scanner.next();

        System.out.print("Genero: ");
        String genero = scanner.next();

        Anime anime = new Anime(nombre, episodios, genero);
        System.out.println("\nInformacion del anime: ");
        anime.mostrarInfo();
    }


    /**
     * 14. Inventario de videojuegos
     * Crear una clase Videojuego con los atributos titulo, plataforma y horasJugadas. Permitir al
     * usuario crear un inventario con varios videojuegos. El programa debe permitir añadir,
     * eliminar y mostrar los videojuegos del inventario
     */
    public void ejercicio14(){
        InventarioVideojuegos inventarioJuegos = new InventarioVideojuegos();

        Videojuego juego1 = new Videojuego("COD", "PS4", "40");
        Videojuego juego2 = new Videojuego("FIFA", "PS4", "799");

        inventarioJuegos.addVideojuego(juego1);
        inventarioJuegos.addVideojuego(juego2);

        System.out.println("Introduce datos videojuego: ");
        System.out.print("Titulo: ");
        String titulo = scanner.next();

        System.out.print("Plataforma: ");
        String plataforma = scanner.next();

        System.out.print("Horas jugadadas: ");
        String horas = scanner.next();

        Videojuego juego3 = new Videojuego(titulo, plataforma, horas);
        inventarioJuegos.addVideojuego(juego3);

        inventarioJuegos.mostrarVideojuegos();
        inventarioJuegos.removeVideojuego(juego1);
        inventarioJuegos.mostrarVideojuegos();

    }

}

File: Ej2_Kotlin/src/main/kotlin/Ejercicios.kt
package org.example
import java.lang.Math.PI
import kotlin.math.sqrt
import kotlin.random.Random


class Ejercicios {

    /**
     * 1. Saludo personalizado
     * Pedir al usuario su nombre y saludarle con un mensaje personalizado. Ejemplo: "¡Hola,
     * Juan!"
     */
    fun ejercicio1(){
        println("** Ejercicio 1 **")
        print("Introduce nombre: ")
        val nombre = readln()
        print("Hola, $nombre")
    }


    /**
     * 2. Número par o impar
     * Pedir un número al usuario e indicar si es par o impar.
     */
    fun ejercicio2(){
        println("** Ejercicio 2 **")
        print("Introduce un numero: ")
        val num = readln().toIntOrNull()
        if (num != null) {
            if (num % 2 == 0) println("Numero $num es par") else println("Numero $num es impar")
        }else{
            println("Debes introducir un valor numerico entero!!")
        }
    }

    /**
     * 3. Cálculo del área de un círculo
     * Pedir al usuario el radio de un círculo y calcular su área (A = π * r^2)
     */
    fun ejercicio3(){
        println("** Ejercicio 3 **")
        print("Introduce el radio de un circulo: ")
        val radio = readln().toFloatOrNull()
        if (radio != null) {
            val area = PI * (radio * radio)
            println("Area de un circulo de $radio de radio: ${"%.2f".format(area)}")
        }else{
            println("Debes introducir un valor numerico!!")
        }
    }

    /**
     * 4. Generador de números aleatorios
     * Generar un número aleatorio entre 1 y 100 y pedir al usuario que adivine qué número es.
     * El programa indicará si el número es mayor o menor al ingresado.
     * Pista: Utilizar la función de generación de números aleatorios y bucles while o do-while.
     */
    fun ejercicio4(){
        println("** Ejercicio 4 **")
        val randNum = Random.nextInt(0,100)
        print("Introduce numero: ")
        var intentos = 0

        do {
            val num = readln().toIntOrNull()
            if (num==null){
                println("El numero debe ser un valor numerico entero!!")
                continue
            }

            intentos++

            when {
                num < randNum -> println("Numero oculto es mayor...")
                randNum < num -> println("Numero oculto es menor...")
            }

        }while (num != randNum)

        println("Has acertado en $intentos intentos!")
    }

    /**
     * 5. Tabla de multiplicar
     * Pedir un número al usuario e imprimir su tabla de multiplicar del 1 al 10.
     */
    fun ejercicio5(){
        println("** Ejercicio 5 **")
        print("Introduce un numero: ")
        val num = readln().toFloatOrNull()

        if (num != null) {
            for (i in 0..<10){
                println("$num x ${i+1} -> ${num*(i+1)}")
            }
        }else{
            println("Debes introducir un valor numerico!!")
        }
    }


    /**
     * 6. Sumar números del 1 al 100
     * Crear un programa que sume todos los números del 1 al 100.
     */
    fun ejercicio6(){
        println("** Ejercicio 6 **")
        var num = 0
        for (i in 0..<100){
            num += (i+1)
        }
        println("Suma total (1-100) -> $num")
    }

    /**
     * 7. Reverso de una cadena
     * Pedir una cadena al usuario e imprimirla al revés.
     */
    fun ejercicio7(){
        println("** Ejercicio 7 **")
        print("Introduce una cadena de texto: ")
        println(readln().reversed())
    }


    /**
     * 8. Contador de vocales
     * Pedir una frase al usuario y contar cuántas vocales (a, e, i, o, u) contiene.
     * 9
     */
    fun ejercicio8(){
        println("** Ejercicio 8 **")
        val vocales = mutableMapOf<Char, Int>(
            'a' to 0,
            'e' to 0,
            'i' to 0,
            'o' to 0,
            'u' to 0
        )

        print("Introduce una frase: ")
        val frase = readln().toLowerCase()

        for (i in frase){
            if (i in vocales.keys){
                vocales[i] = vocales[i]!! + 1
            }
        }

        println("Cantidad de vocales en la frase: $frase")
        for ((vocal, cant) in vocales){
            println("$vocal: $cant")
        }
    }


    /**
     * 9. Números primos
     * Pedir un número al usuario y verificar si es un número primo.
     */
    fun ejercicio9(){
        println("** Ejercicio 9 **")
        print("Introduce un numero: ")
        val num = readln().toIntOrNull()
        if (num != null) {
            if (esPrimo(num)){
                println("Es primo")
            }else{
                println("No es primo")
            }
        }

    }

    private fun esPrimo(num: Int): Boolean{
        if (num < 2) return false
        val sqrt = sqrt(num.toDouble()).toInt()
        for (i in 2..sqrt) {
            if (num % i == 0) {
                return false
            }
        }
        return true
    }


    /**
     * 10. Conversión de temperaturas
     * Crear un programa que convierta una temperatura de grados Celsius a Fahrenheit y
     * viceversa. (C = (F - 32) * 5/9, F = C * 9/5 + 32).
     */
    fun ejercicio10(){
        println("** Ejercicio 10 **")
        print("Que quieres hacer\n1- Celsisus a Fahr\n2- Fahr a Celsius: ")
        val elecc = readln().toIntOrNull()
        if (elecc != null && elecc in 1..2) {
           when (elecc) {
               1 -> celsisuToFah()
               2 -> fahToCelsius()
           }
        }else{
            println("Debes elegir 1,2")
        }

    }

    private fun celsisuToFah(){
        print("Introduce temp en celsius: ")
        val temp = readln().toDoubleOrNull()
        if (temp != null) {
            val res = (temp * 9 / 5) + 32
            print("$temp Celsius en Fahrenheit -> $res")
        }else{
            println("Debes introducir un valor numerico")
        }
    }

    private fun fahToCelsius(){
        print("Introduce temp en Fahrenheit: ")
        val temp = readln().toDoubleOrNull()
        if (temp != null) {
            val res = (temp - 32) * 5/9
            print("$temp Fahrenheit en Celsius -> ${"%.2f".format(res)}")
        }else{
            println("Debes introducir un valor numerico")
        }
    }


    /**
     * 11. Fibonacci
     * Crear un programa que imprima los primeros N números de la secuencia de Fibonacci (el
     * número N lo ingresa el usuario).
     */
    fun ejercicio11(){
        println("** Ejercicio 11 **")
        print("Introduce numero: ")
        val num = readln().toIntOrNull()
        if (num == null) {
            println("Debes introducir un valor numerico")
            return
        }

        var n1 = 0
        var n2 = 1

        println("Los primeros $num numeros de Fibonacci son: ")
        for (i in 1..num){
            print("$n1 ")

            val n3 = n1 + n2
            n1 = n2
            n2 = n3
        }
    }

    /**
     * 12. Inversión de un número
     * Pedir un número entero al usuario e imprimir el número con sus dígitos invertidos.
     * Ejemplo: si el usuario ingresa 1234, el programa debería imprimir 4321.
     */
    fun ejercicio12(){
        println("** Ejercicio 12 **")
        print("Introduce numero para darle la vuelta: ")
        val num = readln().toIntOrNull()
        if (num == null) {
            println("Debes introducir un valor numerico")
            return
        }

        println(num.toString().reversed())
    }


    /**
     * 13. Clase Anime sencilla
     * Crear una clase Anime con los atributos nombre, episodios y género. Luego, permitir al
     * usuario crear un objeto de tipo Anime con estos atributos y mostrar su información.
     */
    fun ejercicio13(){
        println("** Ejercicio 13 **")
        print("Datos creacion clase anime\nIntroduce nombre: ")
        val nombre = readln()

        print("Introduce episodios: ")
        val episodios = readln()

        print("Introduce genero: ")
        val genero = readln()

        val anime = Anime(nombre, episodios, genero)
        println(anime)
    }

    /**
     * 14. Inventario de videojuegos
     * Crear una clase Videojuego con los atributos titulo, plataforma y horasJugadas. Permitir al
     * usuario crear un inventario con varios videojuegos. El programa debe permitir añadir,
     * eliminar y mostrar los videojuegos del inventario.
     */
    fun ejercicio14(){
        println("** Ejercicio 14 **")
        val inventario = InventarioVideojuegos()

        val juego1 = Videojuego("COD", "PS3", 2500)
        val juego2 = Videojuego("F1", "PC", 300)

        val datos = pedirDatos()
        val juego3 = Videojuego(datos[0], datos[1], datos[2].toInt())



        inventario.addJuegos(juego1)
        inventario.addJuegos(juego2)
        inventario.addJuegos(juego3)

        inventario.mostrarJuegos()

        inventario.deleteJuegos(juego1)
        inventario.deleteJuegos(juego2)

        inventario.mostrarJuegos()

    }

    private fun pedirDatos(): List<String>{
        print("Introduce datos juego\nTitulo del juego: ")
        val titulo = readln()
        print("Plataforma del juego: ")
        val plataforma = readln()
        print("Horas jugadas al juego: ")
        val horas = readln().toIntOrNull() ?: 0

        return listOf(titulo, plataforma, horas.toString())
    }
}

